File: JEON-VITY/Assets/Script/Network/Game/NetworkManager.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Runtime.Serialization.Json;
using System.Text;
using UnityEngine;
using UnityEngine.Experimental.UIElements;
using UnityEngine.Networking;
using UnityEngine.SceneManagement;
using UnityEngine.UI;

public class NetworkManager : MonoBehaviour
{

    public List<GameObject> mCharacter;

    public GameObject loadingCanvas; // 로딩중 화면을 보여주는 캔버스 
    public GameObject joystick; // 캐릭터 조종할 조이스틱
    public GameObject buttonA;
    public GameObject buttonB;
    public GameObject handler;
    public GameObject joysticks;
    public GameObject gameoverPop;// 게임종료 팝업
    public GameObject gamequitPop;
    public Text gameoverTimer;
    public GameObject clientCharacter;

    [Header("**groupNavi**")]
    public GameObject voiceChatInput;

    [Header("**groupNavi**")]
    public GameObject groupNaviOpen;
    public Text userA;
    public Text statusA;
    public Text userB;
    public Text statusB;
    public Text userC;
    public Text statusC;
    public Text userD;
    public Text statusD;
    public GameObject groupNaviClose;
    public GameObject btnClose;
    public GameObject btnQuit;
    public int dieCount;

    public void GroupNaviOpen()
    {
        groupNaviOpen.SetActive(true);
        groupNaviClose.SetActive(false);
    }
    public void GroupNaviClose()
    {
        groupNaviOpen.SetActive(false);
        groupNaviClose.SetActive(true);
    }



    [Header("**Socket**")]
    public string host;   // 호스트 주소 
    public int port;  // 포트번호 
    public UdpClient socket; // udp 소켓
    public bool udpSocketReady = false; // 처음 접속한것인지  
    public bool characterSet = false; // 캐릭터가 셋팅이 되었는지
    //public bool move = false; // 캐릭터가 움직이고 있는지 
    public string owner; // 방장이름 
    public string username; // 내 이름 
    public int playerNumber; // 방안에서의 넘버 
    public int group; // 그룹인지 개인전인지 
    public string myPoint; // 나의 시작 포지션 
    public bool isOwner;// 오너인지 아닌지 
    // IPEndPoint는 
    //스스로의 주소(서버의 주소, 즉 IPAddress)를 지정하는 클래스이다
    IPEndPoint epRemote; 
   
    byte[] bytes; // 패킷으로 전송받은 데이터가 담길곳 
    //JSON 
    string roomFilename = "room.json"; // 플레이어 상태 저장 
    string roomPath; // json 파일 저장경로 
    public RoomData roomData = new RoomData(); // 전달할 나의 정보  

    private void Awake()
    {
        Application.targetFrameRate = 60;
        var lm = GameObject.Find("Lobby/LobbyBox/LobbyManager").GetComponent<LobbyManager>();
        username = lm.clientName;
        owner = lm.ownerName;
        playerNumber = lm.playerCount;
        group = lm.groupCount;
        lm.lobbyBox.SetActive(false);
    }

    private void Start()
    {
        if(owner.Equals(username))
        {
            isOwner = true;
        }
        else
        {
            isOwner = false;
        }

        //Debug.Log("cobj" + GameObject.Find(name + "(Clone)").name);
        roomPath = Application.persistentDataPath + "/" + roomFilename; 
        var ps = GameObject.Find("PlayerSpawn").GetComponent<PlayerSpawnManager>();
        //고정프레임
        // Debug.Log("게임시작!");
        // 시작위치 정해주기 
       ps.Spawn(playerNumber, group, username, owner, isOwner);

    //패킷수신 대기
    // IPAddress.Any는 서버자체에 존재하는 IP를 사용
        epRemote = new IPEndPoint(IPAddress.Any, 0);
        bytes = socket.Receive(ref epRemote);
        // 패킷받기 코루틴 시작 
        StartCoroutine(receivePacket());
        //StartCoroutine(LoadTimer());


        // 단체전이면 그룹 네비게이션이 활성화됨 
        if(group == 4)
        {
            GroupNaviOpen();
        }
    }

    public void ConnectToUdpServer()
    {
        if (udpSocketReady)
            return;
        try
        {

            udpSocketReady = true;
            roomData.firstData = true;
            roomData.owner = owner;
            roomData.username = username;
            roomData.group = group.ToString();
            roomData.myPoint = myPoint;
            roomData.playerNumber = playerNumber.ToString();
            roomData.alive = true;
            roomData.winner = false;
            roomData.characterSet = false;
            roomData.move = false;
            // roomData.receivePacket = false;
            roomData.punch = false;
            roomData.throwItem = false;//
            roomData.throwMode = false;//
            roomData.pickup = false;//
            roomData.drop = false;//
            roomData.pX = "";
            roomData.pY = "";
            roomData.pZ = "";
            roomData.rY = "";
            roomData.runX = "";
            roomData.runZ = "";

            RoomSendData();
        }
        catch (Exception e)
        {
            Debug.Log("Socket error : " + e.Message);
        }

        roomData.firstData = false;
        RoomSaveData();
    }



    // 룸정보를 json 데이터로 보낸다.. 
    public void RoomSendData()
    {
        socket = new UdpClient();
        IPEndPoint ep = new IPEndPoint(IPAddress.Parse(host), port);
        socket.Connect(ep);
        // json 데이터의 포장지 느낌   
        RoomWrapper wrapper = new RoomWrapper();
        wrapper.roomData = roomData;
        string room = JsonUtility.ToJson(wrapper, true);
        System.IO.File.WriteAllText(roomPath, room); // json 파일에 작성 
        //string s = System.IO.File.ReadAllText(roomPath);
        //// Json 데이터 보내기 
        //MemoryStream memoryStream = new MemoryStream();
        ////.NET 객체를 JSON으로 변환하고 다시 복원할때 사용한다.  
        //DataContractJsonSerializer serRoom = new DataContractJsonSerializer(typeof(RoomWrapper));
        //serRoom.WriteObject(memoryStream, wrapper);
        //memoryStream.Position = 0;
        //StreamReader sr = new StreamReader(memoryStream);
        var data = Encoding.UTF8.GetBytes(room);
        //  Debug.Log("b" + data.ToString());
        socket.Send(data, data.Length);

    }
   
    // 서버로부터 게임화면 넘어갈때 패킷받기 
    public void ReceivePacket(string data)
    {
        bytes = null;
        try
        {
            JObject obj = JObject.Parse(data);
            string setGame = obj["setGame"].ToString();
            string li = obj["listSize"].ToString();
            bool list = Convert.ToBoolean(li);
            //Debug.Log("listSize" + list);
            // 서버에 있는 json 파일에 내가 속해있는 방의 유저의 정보가 모두 들어왔는지를 판단.
            // 모두 들어와있지 않으면 
            if (!list)
            {
                //Debug.Log("한번더");
                // 한번더 패킷을 받아 업데이트한다. 
                RoomSendData();
                epRemote = new IPEndPoint(IPAddress.Any, 0);
                bytes = socket.Receive(ref epRemote);
            }
            // 모두 들어와있으면 
            if (list)
            {
                socket.Close();
                //  Debug.Log(data);
                JArray playerList = JArray.Parse(setGame);
                {
                    foreach (JObject gobj in playerList.Children<JObject>())
                    {
                        string nm = gobj["name"].ToString();
                        string po = gobj["position"].ToString();
                        // 다른유저 세팅 
                        if (!nm.Equals(username))
                        {
                            var ps = GameObject.Find("PlayerSpawn").GetComponent<PlayerSpawnManager>();
                            ps.OtherSpawn(owner,nm, po, group, isOwner); // 스폰매니저를 실행하여 각자의 위치를 지정한다. 

                           // Debug.Log("다른유저위치1");
                           // Debug.Log(nm);
                           // Debug.Log(po);
                           // Debug.Log(group);
                        }
                    }
                }
                // 코루틴 종료 
                StopCoroutine(receivePacket());
            }
        }
        catch (Exception e)
        {
            Debug.Log("Socket error : " + e.Message);
        }
    }
    // 패킷받기 코루틴 
    private IEnumerator receivePacket()
    {
        while(true)
        {
            if (bytes != null && !roomData.characterSet)
            {
                string data = Encoding.UTF8.GetString(bytes);
                //  Debug.Log("상대접속패킷받기");
                ReceivePacket(data);
            }
            yield return new WaitForSeconds(0.02f);
        }
    }


    public void BackToLobby()
    {
        CloseSocket();
        GameObject.Find("Lobby").transform.Find("LobbyBox").gameObject.SetActive(true);
        var lm = GameObject.Find("Lobby").transform.Find("LobbyBox/LobbyManager").GetComponent<LobbyManager>();
        lm.ComebackLobby();
        SceneManager.LoadScene("ComebackLobby");

    }

    public void BackToRoomOut()
    {
        StartCoroutine(SendUserLose());
        CloseSocket();
        GameObject.Find("Lobby").transform.Find("LobbyBox").gameObject.SetActive(true);
        var lm = GameObject.Find("Lobby").transform.Find("LobbyBox/LobbyManager").GetComponent<LobbyManager>();
        lm.ComebackRoomOut();
        SceneManager.LoadScene("ComebackLobby");
    }

    // 유저의 패배카운트 업데이트하기 
    IEnumerator SendUserLose()
    {
        WWWForm form = new WWWForm();
        form.AddField("username", username);
        using (UnityWebRequest www = UnityWebRequest.Post("http://15.164.101.199/sendlose.php", form))
        {
            yield return www.SendWebRequest();
            if (www.isNetworkError || www.isHttpError)
            {
                Debug.Log(www.error);
                // popupText.text = www.downloadHandler.text;
            }
            else
            {
                //Debug.Log(www.downloadHandler.text);
            }
        }
    }

    private IEnumerator LoadTimer()
    {
        yield return new WaitForSeconds(3f);// 5초뒤
        loadingCanvas.SetActive(false);
        Renderer bA = buttonA.GetComponent<Renderer>();
        //joystick.SetActive(true);
    }

    //jsondata 저장 하기 
    public void RoomSaveData()
    {
        // json 데이터의 포장지 느낌   
        RoomWrapper wrapper = new RoomWrapper();
        wrapper.roomData = roomData;
        string room = JsonUtility.ToJson(wrapper, true);
        System.IO.File.WriteAllText(roomPath, room); // json 파일에 작성 
    }

    public void CloseSocket()
    {
        if (udpSocketReady)
            return;
        socket.Close();
        roomData.throwItem = false;
        roomData.throwMode = false;
        roomData.pickup = false;
        roomData.drop = false;
        udpSocketReady = false;
        roomData.characterSet = false;
        roomData.move = false;
       // roomData.receivePacket = false;
        roomData.punch = false;
        roomData.alive = false;
        roomData.winner = false;
        roomData.owner = "";
        roomData.username = "";
        roomData.myPoint = "";
        roomData.group = "";
        roomData.playerNumber = "";
        roomData.pX = "";
        roomData.pY = "";
        roomData.pZ = "";
        roomData.rY = "";
        roomData.runX = "";
        roomData.runZ = "";
        RoomSaveData();

    }


    // 게임포기하기 팝업 오픈 
    public void QuitGame()
    {
        gamequitPop.SetActive(true);
    }

    // 게임포기
    public void QuitOK()
    {
        gamequitPop.SetActive(false);
        var ch = clientCharacter.GetComponent<Player>(); // 나갈때 내 캐릭터를 죽임
        ch.health = -10;
    }
    // 게임포기팝업 닫기 
    public void QuitCancel()
    {
        gamequitPop.SetActive(false);
    }

    void OnApplicationQuit()
    {
        CloseSocket();
    }

}

File: Vity-server/src/NetworkManager.java
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.JSONValue;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;


public class NetworkManager extends Thread {
	
	int port =9321;
	byte[] data = new byte[65535];
	public DatagramSocket socket;
	DatagramPacket packet;
	JSONParser jsonParser;
	InetAddress address;
	public void run() {
		
		try {
			// 상대방이 연결할수 있도록 udp 소켓생성
			jsonParser = new JSONParser();
			socket = new DatagramSocket(port);
			packet = new DatagramPacket(data, data.length);
			System.out.println("GameRoom접속 대기중 ...");
			//데이터전송받기
			while(true) {
					socket.receive(packet);	
					String str = new String (packet.getData(), packet.getOffset(), packet.getLength(),"Utf-8");
					//System.out.println(str);
						//Thread thread = new MultiGameThread(socket,packet,name);
						//thread.start();
					
				try {
					JSONObject obj = (JSONObject) jsonParser.parse(str);
					JSONObject r = (JSONObject) obj.get("roomData");
					String name = ((String)r.get("username").toString()); // 유저네임
					String owner = ((String)r.get("owner").toString()); // 방장네임
					String myPoint = ((String)r.get("myPoint").toString()); // 포인트
					//String pn = ((String)r.get("playerNumber").toString()); // 포인트
					String gr = ((String)r.get("group").toString()); // 포인트
					//int playerNumber = Integer.parseInt(pn);
					int group = Integer.parseInt(gr);
					boolean firstData = ((boolean)r.get("firstData"));
					boolean characterSet = ((boolean)r.get("characterSet"));
					boolean move = ((boolean)r.get("move"));
					boolean punch = ((boolean)r.get("punch"));
					boolean alive = ((boolean)r.get("alive"));
					boolean winner = ((boolean)r.get("winner"));
					boolean throwMode = ((boolean)r.get("throwMode"));
					boolean throwItem = ((boolean)r.get("throwItem"));
					boolean pickup = ((boolean)r.get("pickup"));
					boolean drop = ((boolean)r.get("drop"));
					//boolean receivePacket = ((boolean)r.get("receivePacket"));
					String px = ((String)r.get("pX").toString()); // 포인트
					String py = ((String)r.get("pY").toString()); // 포인트
					String pz = ((String)r.get("pZ").toString()); // 포인트
					String ry = ((String)r.get("rY").toString()); // 포인트
					String runX = ((String)r.get("runX").toString()); // 포인트
					String runZ = ((String)r.get("runZ").toString());
					Object udpObj = jsonParser.parse(new FileReader("/home/ubuntu/Vity-server/udpRoom.json"));
					JSONObject uob = (JSONObject) udpObj;
					JSONArray udpRoom = (JSONArray) uob.get("udpRoom");
					//String udpJson = JSONValue.toJSONString(uob);
					//System.out.println(uob);
					//System.out.println(udpJson);
					
					
					if(!characterSet) {
						//udp 룸 json 파일을 파싱한다. 
						for(int i = 0; i < udpRoom.size(); i++) {
							JSONObject tmp = (JSONObject) udpRoom.get(i);
							String ow = (String) tmp.get("owner");
							// 오브젝트안에 owner에 내가 속한방의 방장이름이 들어있으면 
							if(ow.equals(owner)) {
								JSONArray pList = (JSONArray)tmp.get("playerList");
								//맨처음 접속시 추가하기
								if(firstData) {
									JSONObject upo = new JSONObject();
									upo.put("name",name);
									//upo.put("address",packet.getAddress().toString());
									//upo.put("port",String.valueOf(packet.getPort()));
									upo.put("position",myPoint);
									upo.put("x",px);
									upo.put("y",py);
									upo.put("z",pz);
									upo.put("ry",ry);
									upo.put("runX",runX);
									upo.put("runZ",runZ);
									upo.put("punch",punch);
									upo.put("alive",alive);
									upo.put("move", move);
									upo.put("winner", winner);
									upo.put("drop", drop);
									upo.put("pickup", pickup);
									upo.put("throwItem", throwItem);
									upo.put("throwMode", throwMode);
									
									//upo.put("move",move);
									pList.add(upo);
									// 방정보를 저장하고있는 json 파일을 업데이트한다.  
									try (FileWriter file = new FileWriter("/home/ubuntu/Vity-server/udpRoom.json")) // json 파일에 문자를 작성한다. 
									{
										file.write(uob.toString()); // jsonobject에 들어있는 문자를 입력 
										file.flush();//stream에 남아있는 데이터를 강제로 내보내는 역할을 한다. 
									}catch(IOException e){e.printStackTrace();}// 파일에 추가하기	
									
									// 방정보를 저장하고있는 json 파일을 업데이트한다.  
									try (FileWriter file = new FileWriter("/home/ubuntu/Vity-server/"+ow+".json")) // json 파일에 문자를 작성한다. 
									{
										file.write(pList.toString()); // jsonobject에 들어있는 문자를 입력 
										file.flush();//stream에 남아있는 데이터를 강제로 내보내는 역할을 한다. 
									}catch(IOException e){e.printStackTrace();}// 파일에 추가하기	
									
									JSONArray nList = new JSONArray();
									// 방정보를 저장하고있는 json 파일을 업데이트한다.  
									try (FileWriter file = new FileWriter("/home/ubuntu/Vity-server/"+ow+"NPC"+".json")) // json 파일에 문자를 작성한다. 
									{
										file.write(nList.toString()); // jsonobject에 들어있는 문자를 입력 
										file.flush();//stream에 남아있는 데이터를 강제로 내보내는 역할을 한다. 
									}catch(IOException e){e.printStackTrace();}// 파일에 추가하기										
									
									JSONArray iList = new JSONArray();
									// 방정보를 저장하고있는 json 파일을 업데이트한다.  
									try (FileWriter file = new FileWriter("/home/ubuntu/Vity-server/item/"+ow+"Item"+".json")) // json 파일에 문자를 작성한다. 
									{
										file.write(iList.toString()); // jsonobject에 들어있는 문자를 입력 
										file.flush();//stream에 남아있는 데이터를 강제로 내보내는 역할을 한다. 
									}catch(IOException e){e.printStackTrace();}// 파일에 추가하기	
									
									// 플레이리스트에 추가된 사람들에게 정보를 업데이트한다.
									//System.out.println(name+" udp connect");
									
								}// 접속한 사람의 정보를 추가한다.
									
								if(group==2) {
									JSONObject setGame = new JSONObject();													
									setGame.put("setGame",pList);
									
									if(pList.size()==2) {
									setGame.put("listSize","true");	
									}else {
										setGame.put("listSize","false");
									}
									try {
										String setGameJson = JSONValue.toJSONString(setGame);
										byte[] byteArr = setGameJson.getBytes("UTF-8");
										DatagramPacket sendPacket = 
												new DatagramPacket(
														byteArr,
														byteArr.length,
														packet.getAddress(),
														packet.getPort());
										socket.send(sendPacket);	
									}catch(Exception e) {
										System.out.println(e.getMessage());
									}	
									
								}
								if(group==4) {
									JSONObject setGame = new JSONObject();													
									setGame.put("setGame",pList);
									
									if(pList.size()==4) {
									setGame.put("listSize","true");	
									}else {
										setGame.put("listSize","false");
									}
									try {
										String setGameJson = JSONValue.toJSONString(setGame);
										byte[] byteArr = setGameJson.getBytes("UTF-8");
										DatagramPacket sendPacket = 
												new DatagramPacket(
														byteArr,
														byteArr.length,
														packet.getAddress(),
														packet.getPort());
										socket.send(sendPacket);	
									}catch(Exception e) {
										System.out.println(e.getMessage());
									}
								}
							}
						}
					}
				}catch(ParseException e) {
					System.out.println(e.getMessage());
				}
				//socket.close();
			}
			
		}catch(Exception e) {
			System.out.println(e.getMessage());
		}
		socket.close();
	}
	
}